docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+
+* Each engine is responsible for persisting engine-specific stuff which needs to be persisted. An obvious thing which needs to be persisted is the syncID and version string for the engine itself, so it can take the necessary action when these change.
+
+* The sync manager will manage keys and pass the relevant key into the engine. Keys for engines should not be exposed to engines other than the key it is for (although currently, in practice, all engines share the same key, but this is an implementation detail.
+
+* The sync manager will pass the relevant part of info/collections to the engine and a token which can be used to access storage. Specific engines never talk to the token server.
+
+# Details
+## Structure definitions
+
+While we use rust struct definitions here, it's important to keep in mind that as mentioned above, we'll need to support the manager being written in something other than rust. As the data is relatively small, it may (or may not) be desirable to use JSON as the serialization format.
+
+    // The main structure passed to a sync implementation. This needs to carry
+    // all state necessary to sync.
+    struct GlobalEngineState {
+        // note: no "declined" or "enabled" here - that's invisible to the engine.
+
+        // the result of info/configuration.
+        config: GlobalConfig,
+
+        // The token server token.
+        storage_token: Vec[u8] (??), // not clear on the type here, but whatever.
+
+        // Info about the collections managed by this engine. In most cases there
+        // will be exactly 1.
+        collections: HashMap<String, CollectionState>
+    }
+
+    // The state for a collection. Most engines will manage a single collection, but
+    // there may be many.
+    struct CollectionState {
+        // The info/collections response for this collection.
+        collection_info: CollectionInfo,
+
+        // The current syncID for the engine, as read from meta/global. May be None.
+        // Engines are responsible for persisting this and taking action when it differs.
+        sync_id: Option<SyncGuid>,
+
+        // The keys to be used for this collection. This may be the default key (ie,
+        // used for all collections) or one specific to the collection, but the
+        // engine doesn't need to know that.
+        keys: KeyBundle,
+
+        // Commands the engine should act upon before syncing.
+        commands: Vec<Command>
+    }
+
+     struct Command {
+        ...
+    }
+
+## APIs
+## Sync Manager
+
+    // Register an engine.
```

Comment:
µnit: ` ```rust `

Reply:
What if this were static, enabled by Cargo `features`? IOW, all available engines are in a static `HashMap`, gated at compile time...and the app passes a list of strings of the engines it wants to enable? (I _think_ that's what @thomcc suggested above, too, but I'm not sure).

So the set of supported engines is fixed, and, when the app does `SyncManager::new(&["bookmarks", "history"])`, or however that's translated over the FFI, this enables them.

Reply:
I've expanded on this significantly - let me know if the new version looks better.